activerecord/test/cases/adapter_test.rb

```diff
@@ -261,6 +283,14 @@ def test_log_invalid_encoding
         assert_not_nil error.message
       end
     end
+
+    private
+
+      def syntax_error_exception_class
+        return Mysql2::Error if defined?(Mysql2)
+        return PG::SyntaxError if defined?(PG)
+        return SQLite3::SQLException if defined?(SQLite3)
```

Comment:
I agree that it feels brittle, and I'm not terribly happy with it. I thought it would be valuable to make sure that the StatementInvalid's `cause` is set to the original exception which caused the error, because I was remembering how StatementInvalid used to store the original exception itself back in 4.2. Looking at 5.x, however, it seems like it just relies on Ruby's nested exceptions now, so I guess we don't need to test this at all.

Anyhow, I couldn't find a convenient way to make `execute` throw a reliable exception which we could test for, so I went with the hard-coded exception classes. Well, that's not quite true. If you do `execute(nil)` it'll throw a TypeError under MySQL and SQLite3, but that's driver-specific behaviour which seems even worse to depend upon, because the dependency is implicit rather than explicit. (Didn't test the PG adapter.)

Sounds like the best approach is to just test for the presence of `cause` and not look any deeper at it. I'll do that for the moment; let me know if you have a better suggestion.

Thanks very much for the feedback!